sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -238,12 +246,11 @@ case class AtomicReplaceTableAsSelectExec(
  */
 case class AppendDataExec(
     table: SupportsWrite,
-    writeOptions: CaseInsensitiveStringMap,
-    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+    writeBuilder: WriteBuilder,
+    query: SparkPlan) extends V2TableWriteExec {
```

Comment:
Since thinking about the impact to `simpleString`, I realized that this is also going to delegate to the write builder for other methods as well, including `equals` and `hashCode`. Since we can't rely on the behavior of the write builder's `equals` and `hashCode` methods, I don't think it the builder should be used as an argument to the plan case classes.

I think that also makes sense: case classes are algebraic types, so we shouldn't include objects in their definitions that don't behave like algebraic types.